src/Instrumentation/ExtRdKafka/src/ExtRdKafkaInstrumentation.php

```diff
@@ -0,0 +1,131 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OpenTelemetry\Contrib\Instrumentation\ExtRdKafka;
+
+use Composer\InstalledVersions;
+use OpenTelemetry\API\Globals;
+use OpenTelemetry\API\Instrumentation\CachedInstrumentation;
+use OpenTelemetry\API\Trace\Propagation\TraceContextPropagator;
+use OpenTelemetry\API\Trace\Span;
+use OpenTelemetry\API\Trace\SpanKind;
+use OpenTelemetry\Context\Context;
+use function OpenTelemetry\Instrumentation\hook;
+
+use OpenTelemetry\SemConv\TraceAttributes;
+
+use RdKafka\KafkaConsumer;
+use RdKafka\Message;
+use RdKafka\ProducerTopic;
+
+use Throwable;
+
+class ExtRdKafkaInstrumentation
+{
+    public const NAME = 'ext_rdkafka';
+    public static function register(): void
+    {
+        $instrumentation = new CachedInstrumentation(
+            'io.opentelemetry.contrib.php.ext_rdkafka',
+            InstalledVersions::getVersion('open-telemetry/opentelemetry-auto-ext-rdkafka'),
+            TraceAttributes::SCHEMA_URL,
+        );
+
+        // Start root span and propagate parent if it exists in headers, for each message consumed
+        self::addConsumeHooks($instrumentation);
+        // End root span on offset commit
+        self::addCommitHooks('commit');
+        self::addCommitHooks('commitAsync');
+        // Context propagation for outbound messages
+        self::addProductionHooks();
+    }
+
+    private static function addCommitHooks($functionName)
+    {
+        hook(
+            KafkaConsumer::class,
+            $functionName,
+            null,
+            static function () {
+                $scope = Context::storage()->scope();
+                if ($scope === null) {
+                    return;
+                }
+                $scope->detach();
+                $span = Span::fromContext($scope->context());
+                $span->end();
+            }
+        );
+    }
+
+    private static function addProductionHooks()
+    {
+        hook(
+            ProducerTopic::class,
+            'producev',
+            static function (
+                ProducerTopic $exchange,
+                array $params
+            ): array {
+                // Headers are the 5th argument for the producev function
+                $carrier = [];
+                TraceContextPropagator::getInstance()->inject($carrier);
+                $propagator = Globals::propagator();
+                $propagator->inject($carrier);
+
+                if (array_key_exists('traceparent', $carrier)) {
+                    $params[4]['traceparent'] = $carrier['traceparent'];
+                }
+
+                return $params;
```

Comment:
A code suggestion on how the span generated at this moment could be composed when producing a new message.

```suggestion
               $builder = $instrumentation
                   ->tracer()
                   ->spanBuilder('publish') // topicName + publish, example foo publish
                   ->setSpanKind(SpanKind::KIND_PRODUCER)
                   ->setAttribute(TraceAttributes::CODE_FUNCTION, $function)
                   ->setAttribute(TraceAttributes::CODE_NAMESPACE, $class)
                   ->setAttribute(TraceAttributes::CODE_FILEPATH, $filename)
                   ->setAttribute(TraceAttributes::CODE_LINENO, $lineno)
                   ->setAttribute(TraceAttributes::MESSAGING_SYSTEM, 'kafka')
                   ->setAttribute(TraceAttributes::MESSAGING_OPERATION, 'publish')
               $parent = Context::getCurrent();
               
               $span = $builder
                    ->setParent($parent)
                    ->startSpan();
               
               // Kafka message key
               if (array_key_exists(3, $params)) {
                     $span->setAttribute(TraceAttributes::MESSAGING_KAFKA_MESSAGE_KEY, $params[3]);
               }
               
               // Headers are the 5th argument for the producev function
               if (sizeof($params) >= 5 && array_key_exists(4, $params)) {
                   $headers = $params[4];
                   
                   $propagator = Globals::propagator();
                   $propagator->inject($headers, ArrayAccessGetterSetter::getInstance(), $context);
               }
               
                return $params;
```

Reply:
Just trying to work this one out (apologies, still getting used to OTEL in PHP and in general, so terminology may be a bit off).

So this adds a span for the actual production of the message (I like it). I can't see it actually activate that span though, guessing I'll want to do that? I'm assuming from there, I'll need to close the span off in the `post` hook right? (which isn't yet there, but fine to add). Otherwise I'll get an _unclosed span_ error (OWTTE)

Also, I see that the `$context` variable is not actually defined, when injecting the context into the headers (appreciate this was probably written on GitHub and not in an IDE). Should I use the same context that I'm using for the parent of this span, currently defined as `$parent`, or use the context of the newly created span (which I'm assuming I'll need to activate prior). Guessing the newly created span, but wanted to make sure.

Cheers :)

Reply:
This is correct, in this example that I added as a suggestion, the span should be closed in the `post hook`.
Regarding the written code, I ended up writing directly through Github and there ended up being confusion when declaring the variables, but below is what the `pre hook` would be like.

```php
$builder = $instrumentation
    ->tracer()
    ->spanBuilder('publish') // topicName + publish, example foo publish
    ->setSpanKind(SpanKind::KIND_PRODUCER)
    ->setAttribute(TraceAttributes::CODE_FUNCTION, $function)
    ->setAttribute(TraceAttributes::CODE_NAMESPACE, $class)
    ->setAttribute(TraceAttributes::CODE_FILEPATH, $filename)
    ->setAttribute(TraceAttributes::CODE_LINENO, $lineno)
    ->setAttribute(TraceAttributes::MESSAGING_SYSTEM, 'kafka')
    ->setAttribute(TraceAttributes::MESSAGING_OPERATION, 'publish')

$parent = Context::getCurrent();

$span = $builder
    ->setParent($parent)
    ->startSpan();

$context = $span->storeInContext($parent);

// Kafka message key
if (array_key_exists(3, $params)) {
    $span->setAttribute(TraceAttributes::MESSAGING_KAFKA_MESSAGE_KEY, $params[3]);
}

// Headers are the 5th argument for the producev function
if (sizeof($params) >= 5 && array_key_exists(4, $params)) {
    $headers = $params[4];
    $propagator = Globals::propagator();
    $propagator->inject($headers, ArrayAccessGetterSetter::getInstance(), $context);
}

Context::storage()->attach($context);

return $params;
```

Reply:
I'm sorry, now your initial approach is clear to me. See if the code below helps you with this.

```php
$carrier = [];
$propagator = Globals::propagator();
$propagator->inject($carrier);

// Headers are the 5th argument for the producev function
$params[4] = isset($params[4]) ? array_merge($params[4], $carrier) : $carrier;

return $params;
```

Reply:
All good, I've ended up going for the first one anyway. So it's now creating a span for the lifecycle for the produce method, and still works end to end.

Had to modify the tests to expect more spans, but that's all fine, all passing